### PR TITLE
TCA9548A : initial implementation

### PR DESCRIPTION
**Pull request by g-vidal**
Signed-off-by: g-vidal <gerard.vidal@ens-lyon.fr>

Propositon of functions to activate /close i2c channels from TCA9548A multiplexer. It enables to handle sensors with the same address on different buses, or simply extend the number of available i2c buses.

inspired from #413

### CODE REVIEW

**Comment by g-vidal** [comment]
Sorry
I forgot to check compilation in my example I am using aother sensor (bmp280)  and the include is not correct... I am on it.
Best

**Comment by g-vidal** [comment]
@jontrulson Hi is my request correct or should I do something else ? I have to comment the call to external sensors (which is the purpose of the TCA9548A) but I have left them in the code to help give indications to the users.
Best

**Comment by g-vidal** [comment]
OK
 I am on it  I am merging #413and making necessary amendments.

**Comment by g-vidal** [comment]
Hi @Propanu  @jontrulson @pylbert @malikabhi05, after some trouble in the rebase... (git beginner sorry) I think that things may be correct now  I have integrated most of the code from #413.
Sory for the delay
Thanks

File: examples/c++/tca9548a.cxx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  /*cout << "Temperature 1 : " << sensor1->getTemp()
+       << endl;
+  cout << "Temperature 2 : " << sensor2->getTemp()
+            << endl;
+  cout << "Exiting..." << endl;
+   
+  delete sensor1;
+  delete sensor2;*/
+  multiplex->disableAllPorts();
+
```
Comment (jontrulson) [inline examples/c++/tca9548a.cxx]: Multiplex still needs to be deleted here, ie:

`delete multiplex;`

File: examples/c++/tca9548a.cxx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  // Instantiate a MCP9808 instance using multiplex C4 i2c bus and MCP9808 default address
+  //upm::MCP9808 *sensor2 = new upm::MCP9808(0);
+  //close com with channel 2
+  multiplex->disableAllPorts();
+  //*****************************************************
+  // Open 2 i2c bus on multiplexer channel 4 and channel2
+  //*****************************************************
+  multiplex->setPort(4,multiplex->ENABLED,multiplex->EXCLUSIVE);
+  multiplex->setPort(2,multiplex->ENABLED,multiplex->INCLUSIVE);
+  // we show temperature
```
Comment (jontrulson) [inline examples/c++/tca9548a.cxx]: Could we get rid of these commented out parts?  Or, at least use a consistent commenting style, like '//'?

File: examples/c++/tca9548a.cxx [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  //**************************************
+  // Open i2c bus on multiplexer channel 4
+  //**************************************
+  multiplex->setPort(4,multiplex->ENABLED,multiplex->EXCLUSIVE);
+  // Instantiate a MCP9808 instance using multiplex C4 i2c bus and MCP9808 default address
+  //upm::MCP9808 *sensor1 = new upm::MCP9808(0);
+  // we show temperature
+  //cout << "Temperature: " << sensor1->getTemp()
+  //     << endl;
+  //cout << "Exiting..." << endl;
```
Comment (jontrulson) [inline examples/c++/tca9548a.cxx]: Could we get rid of these commented out parts?

File: examples/javascript/tca9548.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
+/*jslint node:true, vars:true, bitwise:true, unparam:true */
+/*jshint unused:true */
```
Comment (jontrulson) [inline examples/javascript/tca9548.js]: These two lines are no longer needed, they can be removed.

File: examples/c++/tca9548a.cxx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+ * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <unistd.h>
+#include <iostream>
+#include <signal.h>
+
+//#include "../mcp9808/mcp9808.hpp"
```
Comment (jontrulson) [inline examples/c++/tca9548a.cxx]: Could probably remove this?  Even if you wanted to use this sensor, the current method of handling dependencies would handle this for you.  You would add "mcp9808" to the argument list to upm_module_init(), so in your example code you would just use:

`#include "mcp9808.hpp"`

and in your driver CMakeLists.txt file, your call to upm_module init() would look like:

`upm_module_init(mraa mcp9808)`

And the include paths and library dependencies would be handled for you.

Comment (g-vidal) [inline examples/c++/tca9548a.cxx]: Done

File: examples/c++/tca9548a.cxx (hunk 3 above)
Comment (g-vidal) [inline examples/c++/tca9548a.cxx]: Done

File: examples/c++/tca9548a.cxx (hunk 2 above)
Comment (g-vidal) [inline examples/c++/tca9548a.cxx]: Done

File: examples/c++/tca9548a.cxx (hunk 1 above)
Comment (g-vidal) [inline examples/c++/tca9548a.cxx]: Done

File: examples/javascript/tca9548.js (hunk 4 above)
Comment (g-vidal) [inline examples/javascript/tca9548.js]: Done

**Comment by jontrulson** [comment]
Hi @g-vidal, did you forget to push your new fixed, squashed commit?  I don't see the actual changes you say you made...

File: examples/c++/tca9548a.cxx (hunk 5 above)
Comment (jontrulson) [inline examples/c++/tca9548a.cxx]: Sorry, I missed this one.  Ideally, it should not depend on any sensor at all, as this example is meant to be generic, and it's quite ok if it looks "empty" by not actually talking to another sensor device.

**Comment by g-vidal** [comment]
Hi @jontrulson 
This time things look better.... examples are homogenous as requested.
I hope we are done.
Thanks for your patience

**Comment by jontrulson** [comment]
Merged, thanks!

**Comment by g-vidal** [comment]
@Propanu @jontrulson 
Tanks for merging
I did not know howto incorporate it but I called references to #413 that could be closed now 
Thanks

**Pull request closed without merging**